sys/kern/kmem.c

```diff
@@ -28,8 +30,26 @@ static void kick_swapper(void) {
 vaddr_t kva_alloc(size_t size) {
   assert(page_aligned_p(size));
   vmem_addr_t start;
-  if (vmem_alloc(kvspace, size, &start, M_NOGROW))
-    return 0;
+
+  bool restarted = false;
+
+retry:
+  if (vmem_alloc(kvspace, size, &start, M_NOGROW)) {
+    if (restarted)
+      return 0;
```

Comment:
I think this can lead to `kva_alloc()` returning 0 even if `kvspace` can be grown.
Consider what happens if another thread calls `kva_alloc()` right after we call `vmem_add()`, but before we retry the call to `vmem_alloc()`. The other thread can take all the space we just added, so our second attempt to `vmem_alloc()` will fail. I think we should just remove the `restarted` variable altogether and keep trying until we succeed.

Reply:
done